File: csharp/dms-service-sdk-sample/Client/PaymentClientForShop.cs
using System.Text;
using System.Net;
using Newtonsoft.Json.Linq;
using Nethereum.Signer;
using Nethereum.ABI;
using Org.BouncyCastle.Crypto.Digests;

namespace Dms.Service.Sdk.Client;

using Utils;
using Types;

public class PaymentClientForShop(NetWorkType network, string privateKey, string shopId) : Client(network)
{
    private readonly EthECKey _keyPair = new(privateKey);
    public string Address => _keyPair.GetPublicAddress();
    public string ShopId { get; } = shopId;

    // region Payment
    public static byte[] GetLoyaltyCancelPaymentMessage(
        string paymentId,
        string purchaseId,
        string account,
        long nonce,
        long chainId
    )
    {
        var abiEncode = new ABIEncode();
        var encodedBytes = abiEncode.GetABIEncoded(
            new ABIValue("bytes32", CommonUtils.ConvertHexStringToByte(paymentId)),
            new ABIValue("string", purchaseId),
            new ABIValue("address", account),
            new ABIValue("uint256", chainId),
            new ABIValue("uint256", nonce)
        );
        var message = new byte[32];
        var digest = new KeccakDigest(256);
        digest.BlockUpdate(encodedBytes, 0, encodedBytes.Length);
        digest.DoFinal(message, 0);
        return message;
    }

    public async Task<PaymentTaskItemShort> ApproveCancelPayment(
        string paymentId,
        string purchaseId,
        bool approval
    ) {
        var account = this.Address;
        var nonce = await this.GetLedgerNonceOf(account);
        var message = GetLoyaltyCancelPaymentMessage(
            paymentId,
            purchaseId,
            account,
            nonce,
            await this.GetChainId()
        );
        var signature = CommonUtils.SignMessage(this._keyPair, message);
        
        var body = new StringContent(new JObject
        {
            { "paymentId", paymentId },
            { "approval", approval },
            { "signature", signature }
        }.ToString(), Encoding.UTF8, "application/json");

        var response = await PostAsync($"{RelayEndpoint}/v2/payment/cancel/approval", body);

        return PaymentTaskItemShort.FromJObject(ParseResponseToJObject(response));
    }
}

File: java/sample/src/main/java/org/example/PaymentClientForShop.java
package org.example;

import org.kios.service.sdk.client.PaymentClient;
import org.kios.service.sdk.data.NetWorkType;
import org.kios.service.sdk.data.payment.PaymentTaskItemShort;
import org.kios.service.sdk.utils.CommonUtils;
import org.json.JSONObject;
import org.web3j.abi.TypeEncoder;
import org.web3j.abi.datatypes.Address;
import org.web3j.abi.datatypes.DynamicStruct;
import org.web3j.abi.datatypes.Utf8String;
import org.web3j.abi.datatypes.generated.Bytes32;
import org.web3j.abi.datatypes.generated.Uint256;
import org.web3j.crypto.Hash;
import org.web3j.utils.Numeric;

import java.io.OutputStream;
import java.net.HttpURLConnection;
import java.net.URI;
import java.nio.charset.StandardCharsets;

public class PaymentClientForShop extends PaymentClient {
    protected String shopId;

    public PaymentClientForShop(NetWorkType network, String privateKey, String shopId) {
        super(network, privateKey);

        this.shopId = shopId;
    }

    public static byte[] getLoyaltyCancelPaymentMessage(
            String paymentId,
            String purchaseId,
            String account,
            long nonce,
            long chainId
    ) {
        String value = TypeEncoder.encode(
                new DynamicStruct(
                        new Bytes32(Numeric.hexStringToByteArray(paymentId)),
                        new Utf8String(purchaseId),
                        new Address(account),
                        new Uint256(chainId),
                        new Uint256(nonce)
                )
        );
        return Hash.sha3(Numeric.hexStringToByteArray(value));
    }

    public String getShopId() {
        return this.shopId;
    }

    public PaymentTaskItemShort approveCancelPayment(
            String paymentId,
            String purchaseId,
            Boolean approval
    ) throws Exception {
        String account = this.credentials.getAddress();
        long nonce = this.getLedgerNonceOf(account);
        byte[] message = getLoyaltyCancelPaymentMessage(
                paymentId,
                purchaseId,
                account,
                nonce,
                this.getChainId()
        );
        String signature = CommonUtils.signMessage(this.credentials.getEcKeyPair(), message);
        URI uri = new URI(String.format("%s/v2/payment/cancel/approval", relayEndpoint));
        HttpURLConnection conn = getHttpURLConnection(uri, "POST");

        JSONObject body = new JSONObject();
        body.put("paymentId", paymentId);
        body.put("approval", approval);
        body.put("signature", signature);

        try (OutputStream output = conn.getOutputStream()) {
            output.write(body.toString().getBytes(StandardCharsets.UTF_8));
        }

        return PaymentTaskItemShort.fromJSONObject(getJSONObjectResponse(conn));
    }
}
